time/src/time.cpp

```diff
@@ -0,0 +1,559 @@
+    {
+        seconds = static_cast<int64_t>(microseconds / MICROSECONDS_PER_SECOND);
+        microseconds = fmod(microseconds, MICROSECONDS_PER_SECOND);
+    }
+
+    return createDurationFromTimespec(L, {seconds, static_cast<int32_t>(microseconds * NANOSECONDS_PER_MICROSECOND)});
+}
+
+int lua_milliseconds(lua_State* L)
+{
+    double milliseconds = luaL_checknumber(L, 1);
+    if (milliseconds < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
+
+    int64_t seconds = 0;
+    if (milliseconds > MILLISECONDS_PER_SECOND)
+    {
+        seconds = static_cast<int64_t>(milliseconds / MILLISECONDS_PER_SECOND);
+        milliseconds = fmod(milliseconds, MILLISECONDS_PER_SECOND);
+    }
+
+    return createDurationFromTimespec(L, {seconds, static_cast<int32_t>(milliseconds * NANOSECONDS_PER_MILLISECOND)});
+}
+
+int lua_seconds(lua_State* L)
+{
+    double seconds = luaL_checknumber(L, 1);
+    if (seconds < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
+
+    return createDurationFromSeconds(L, seconds);
+}
+
+int lua_minutes(lua_State* L)
+{
+    double minutes = luaL_checknumber(L, 1);
+    if (minutes < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
+
+    return createDurationFromSeconds(L, minutes * SECONDS_PER_MINUTE);
+}
+
+int lua_hours(lua_State* L)
+{
+    double hours = luaL_checknumber(L, 1);
+    if (hours < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
+
+    // hours can still overflow
+    if (hours > (pow(2, 63) / SECONDS_PER_HOUR))
+    {
+        luaL_error(L, "duration is too large");
+        return 0;
+    }
+
+    return createDurationFromSeconds(L, hours * SECONDS_PER_HOUR);
+}
+
+int lua_days(lua_State* L)
+{
+    double days = luaL_checknumber(L, 1);
+    if (days < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
+
+    // account for overflow
+    if (days > (pow(2, 63) / SECONDS_PER_DAY))
+    {
+        luaL_error(L, "duration is too large");
+        return 0;
+    }
+
+    return createDurationFromSeconds(L, days * SECONDS_PER_DAY);
+}
+
+int lua_weeks(lua_State* L)
+{
+    double weeks = luaL_checknumber(L, 1);
+    if (weeks < 0)
+    {
+        luaL_error(L, "duration cannot be negative");
+        return 0;
+    }
+
+    // account for overflow
+    if (weeks > (pow(2, 63) / SECONDS_PER_WEEK))
+    {
+        luaL_error(L, "duration is too large");
+        return 0;
+    }
+
+    return createDurationFromSeconds(L, weeks * SECONDS_PER_WEEK);
+}
+
+} // namespace duration
+
+namespace libtime
+{
+int lua_now(lua_State* L)
+{
+    uv_timespec64_t now;
+
+    int status = uv_clock_gettime(UV_CLOCK_MONOTONIC, &now);
+    assert(status == 0);
+
+    uv_timespec64_t* timespec = static_cast<uv_timespec64_t*>(lua_newuserdatatagged(L, sizeof(uv_timespec64_t), 1));
```

Comment:
define the tag in a constant in this file, we will refactor all those constants into a runtime header later, e.g.
`kInstantTag`